### DISCUSSION TITLE
Arabic words are reversed

### DISCUSSION THREAD

**Discussion by 3mora2** [pymupdf/PyMuPDF#1052] [Looking for help] [answered]
Hey 
thanks for great work but in Arabic words are reversed
Like: 
'العامة والاسكان'
''ناكسالاو ةماعلا''
there way to be correct?

**Comment by JorjMcKie** [comment, marked as answer]
I presume you refer to text output on a PDF?
If so, try using the ``TextWriter`` class: methods `TextWriter.append()` and `TextWriter.fill_textbox()` have the parameter `right_to_left`.
Set this to `True` and it **_should_** work.
The older text output methods `page.insert_text` and friends **_do not support this_**.

**Reply by 3mora2** [reply]
thanks for reply, i ask about page.getText not about TextWrite
i can use many ways to reverse text but some letter shouldn't reverse,
so it must to be by lib

**Reply by JorjMcKie** [reply]
ah, ok.
Try to extract text via `page.get_text("dict")`. On the line level (line-subdictionaries) of the returned dict you will find the key `"dir"` which contains the writing direction. It is the unit vector ``(cosine, sine)`` of the angle of the text with the x-axis viewed from left to right.
For horizontal left-to-right you will therefore find ``(1, 0)``, and for right-to-left you should see ``(-1, 0)``, on which you can base your code's decision whether to revert the extracted text.

**Reply by 3mora2** [reply]
Yes is very good, but still some letter when write from Left to right, then reverse ,don't return to its origin 
Like : 'تللايفات'
reverse to : 'تافياللت'
Origin : 'تافيلالت'

**Reply by JorjMcKie** [reply]
> still some letter when write from Left to right, then reverse ,don't return to its origin

? don't understand.
You mean the ``dir`` is not correctly set, or what?

**Reply by JorjMcKie** [reply]
Thre always is the possibility, that the PDF is coded in a chaotic way: I have seen example, where everything looked fine in a PDF viewer, but the extract was completely out of order: `PFyuPMD` instead of `PyMuPDF`.
You can always use `page.get_text("rawdict")` which gives you the ccordinates of every single character. This makes you independent from everything: sort the character coordinates by y- and then by x-coordinates.

**Reply by 3mora2** [reply]
> > still some letter when write from Left to right, then reverse ,don't return to its origin
> 
> ? don't understand.
> You mean the `dir` is not correctly set, or what?

No, it set correctly
but some letter like 'الا' , i think some from Left to right and other from right to left

**Reply by 3mora2** [reply]
> Thre always is the possibility, that the PDF is coded in a chaotic way: I have seen example, where everything looked fine in a PDF viewer, but the extract was completely out of order: `PFyuPMD` instead of `PyMuPDF`.
> You can always use `page.get_text("rawdict")` which gives you the ccordinates of every single character. This makes you independent from everything: sort the character coordinates by y- and then by x-coordinates.

Yes, it the best to give every single character
Thank you very much 💝

**Reply by 3mora2** [reply]
can ask about format which open arabic file?
